### PR TITLE
add implement dmi interface

### PR DESCRIPTION
**Pull request by cl2017**
What type of PR is this?
/kind feature

What this PR does / why we need it:
add implement dmi interface

Which issue(s) this PR fixes:
NONE

Special notes for your reviewer:

### CODE REVIEW

File: mappers/modbus/model-modbus.yaml [hunk 1]
```diff
@@ -0,0 +1,21 @@
+apiVersion: devices.kubeedge.io/v1alpha2
+kind: DeviceModel
+metadata:
+  name: sensor-tag-model
+  namespace: default
+spec:
```
Comment (RyanZhaoXB) [inline mappers/modbus/model-modbus.yaml:6]: add `protocol: modbus` upon `properties`

File: pkg/apis/dmi/v1/api.proto [hunk 2]
```diff
@@ -0,0 +1,324 @@
+syntax = "proto3";
```
Comment (RyanZhaoXB) [inline pkg/apis/dmi/v1/api.proto:1]: after the pr(https://github.com/kubeedge/kubeedge/pull/4013) is merged, we can remove this `pkg/apis/dmi` directory and import the package in kubeedge directly

File: pkg/driver/modbus/client.go [hunk 3]
```diff
@@ -14,17 +14,17 @@ See the License for the specific language governing permissions and
...
 
-package driver
+package modbus
 
 import (
 	"errors"
 	"sync"
 	"time"
 
+	"github.com/kubeedge/mappers-go/pkg/common"
```
Comment (RyanZhaoXB) [inline pkg/driver/modbus/client.go:24]: the order of import packages should be optimized.

File: pkg/grpcserver/device.go [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+	if err != nil {
+		return nil, fmt.Errorf("parse device %s instance failed, err: %s", device.Name, err)
+	}
+	deviceInstance.PProtocol = protocol
+
+	s.devPanel.UpdateDev(&model, deviceInstance, &protocol)
+
+	// TODO need edgecore publish?
+	// publish device twin to mqtt
+	//topic := dtcommon.DeviceETPrefix + device.Name + dtcommon.DeviceETUpdatedSuffix
```
Comment (RyanZhaoXB) [inline pkg/grpcserver/device.go:56]: remove the unused code

File: pkg/grpcserver/upstream.go [hunk 5]
```diff
@@ -0,0 +1,60 @@
+package grpcserver
+
+import (
+	"context"
+	"errors"
+
+	dmiapi "github.com/kubeedge/mappers-go/pkg/apis/dmi/v1"
```
Comment (RyanZhaoXB) [inline pkg/grpcserver/upstream.go:7]: the order of the import packages should be optimized

File: pkg/util/grpcclient/register.go [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+import (
+	"context"
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/kubeedge/mappers-go/config"
+	dmiapi "github.com/kubeedge/mappers-go/pkg/apis/dmi/v1"
+	"github.com/kubeedge/mappers-go/pkg/common"
+	"google.golang.org/grpc"
```
Comment (RyanZhaoXB) [inline pkg/util/grpcclient/register.go:12]: the order of the import packages should be optimized

File: pkg/util/grpcclient/register.go [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+)
+
+// RegisterMapper if withData is true, edgecore will send device and model list.
+func RegisterMapper(cfg *config.Config, withData bool) ([]*dmiapi.Device, []*dmiapi.DeviceModel, error) {
+	// 连接grpc服务器
```
Comment (RyanZhaoXB) [inline pkg/util/grpcclient/register.go:17]: please comment with English

Comment (RyanZhaoXB) [reply pkg/util/grpcclient/register.go:17]: or the Chinese comments in this file can be all removed.

File: pkg/util/grpcclient/register.go [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+	// 初始化Greeter服务客户端
+	c := dmiapi.NewDeviceManagerServiceClient(conn)
+
+	// 初始化上下文，设置请求超时时间为1秒
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	// 延迟关闭请求会话
+	defer cancel()
+
+	// 调用SayHello接口，发送一条消息
+	fmt.Println("======send grpc to edgecore for mapperRegister")
```
Comment (RyanZhaoXB) [inline pkg/util/grpcclient/register.go:47]: this log is for test. we can remove it.

File: pkg/util/grpcclient/register.go [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+			Protocol:   cfg.Common.Protocol,
+			Address:    []byte(cfg.GrpcServer.SocketPath),
+			State:      common.DEVSTOK,
+		},
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+	fmt.Println("======send grpc to edgecore for mapperRegister finished")
+	fmt.Printf("======register response: %+v", resp)
```
Comment (RyanZhaoXB) [inline pkg/util/grpcclient/register.go:63]: this log is for test. we can remove it.

File: pkg/util/grpcclient/report.go [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+					return nil, err
+				}
+				return net.DialUnix("unix", nil, unixAddress)
+			},
+		),
+	)
+	if err != nil {
+		return fmt.Errorf("did not connect: %v", err)
+	}
+	// 延迟关闭连接
```
Comment (RyanZhaoXB) [inline pkg/util/grpcclient/report.go:32]: the Chinese comments in this file can be all removed.

File: pkg/util/parse/grpc.go [hunk 11]
```diff
@@ -0,0 +1,346 @@
...
+				}
+				res.Properties = append(res.Properties, cur)
+			}
+		}
+	}
+	return res
+}
+
+func buildPropertyVisitorsFromGrpc(device *dmiapi.Device) []common.PropertyVisitor {
+	klog.Infof("##########################")
```
Comment (RyanZhaoXB) [inline pkg/util/parse/grpc.go:184]: these log in this file is for test. we can remove them.

File: pkg/util/parse/grpc.go [hunk 12]
```diff
@@ -0,0 +1,346 @@
...
+		cur := common.PropertyVisitor{
+			Name:          pptv.PropertyName,
+			PropertyName:  pptv.PropertyName,
+			ModelName:     device.Spec.DeviceModelReference,
+			CollectCycle:  pptv.GetCollectCycle(),
+			ReportCycle:   pptv.GetReportCycle(),
+			Protocol:      protocolName,
+			VisitorConfig: visitorConfig,
+		}
+		klog.Infof("cur: %+v", cur)
```
Comment (RyanZhaoXB) [inline pkg/util/parse/grpc.go:239]: remove this log.

File: pkg/util/parse/grpc.go [hunk 13]
```diff
@@ -0,0 +1,346 @@
...
+		res = append(res, cur)
+		klog.Infof("##########################")
+	}
+	klog.Infof("res: %+v", res)
```
Comment (RyanZhaoXB) [inline pkg/util/parse/grpc.go:243]: remove this log.

File: pkg/util/parse/grpc.go [hunk 14]
```diff
@@ -0,0 +1,346 @@
...
+		PropertyVisitors: buildPropertyVisitorsFromGrpc(device),
+	}
+	propertyVisitorMap := make(map[string]common.PropertyVisitor)
+	for i := 0; i < len(instance.PropertyVisitors); i++ {
+		if commonModel == nil {
+			klog.Errorf("commonModel == nil")
+			continue
+		}
+
+		klog.Infof("======commonmodel property len: %d", len(commonModel.Properties))
```
Comment (RyanZhaoXB) [inline pkg/util/parse/grpc.go:322]: remove this log.

File: pkg/util/parse/grpc.go [hunk 15]
```diff
@@ -0,0 +1,346 @@
...
+		for _, property := range commonModel.Properties {
+			if property.Name == instance.PropertyVisitors[i].PropertyName {
+				klog.Infof("======common property %s equal, set instance propertyVisitor PProperty", property.Name)
```
Comment (RyanZhaoXB) [inline pkg/util/parse/grpc.go:325]: remove this log.

File: pkg/util/parse/grpc.go [hunk 16]
```diff
@@ -0,0 +1,346 @@
...
+				instance.PropertyVisitors[i].PProperty = property
+				break
+			}
+		}
+		klog.Infof("======set propertyVisitorMap property, key: %s, value: %+v", instance.PropertyVisitors[i].PProperty.Name, instance.PropertyVisitors[i])
```
Comment (RyanZhaoXB) [inline pkg/util/parse/grpc.go:330]: remove this log.

File: pkg/util/parse/grpc.go [hunk 17]
```diff
@@ -0,0 +1,346 @@
...
+		propertyVisitorMap[instance.PropertyVisitors[i].PProperty.Name] = instance.PropertyVisitors[i]
+	}
+	for i := 0; i < len(instance.Twins); i++ {
+		klog.Infof("====== current instance twin property name %s", instance.Twins[i].PropertyName)
```
Comment (RyanZhaoXB) [inline pkg/util/parse/grpc.go:334]: remove this log.

File: pkg/util/parse/grpc.go [hunk 18]
```diff
@@ -0,0 +1,346 @@
...
+		if v, ok := propertyVisitorMap[instance.Twins[i].PropertyName]; ok {
+			klog.Infof("====== current instance twin property set visitor %+v", v)
```
Comment (RyanZhaoXB) [inline pkg/util/parse/grpc.go:336]: remove this log.

File: pkg/util/parse/parse.go [hunk 19]
```diff
@@ -0,0 +1,406 @@
...
+	}
+	deviceModelList, err := httpclient.GetDeviceModelList(cfg.DevInit.MetaServer.Addr, cfg.DevInit.MetaServer.Namespace)
+	if err != nil {
+		return err
+	}
+	if len(deviceModelList) == 0 {
+		return ErrEmptyData
+	}
+	// TODO test only
+	//deviceList := tmpDevices
```
Comment (RyanZhaoXB) [inline pkg/util/parse/parse.go:335]: remove the unused code

File: pkg/util/parse/parse.go [hunk 20]
```diff
@@ -0,0 +1,406 @@
...
+	if err != nil {
+		return err
+	}
+
+	if len(deviceList) == 0 || len(deviceModelList) == 0 {
+		return ErrEmptyData
+	}
+	modelMap := make(map[string]common.DeviceModel)
+	for _, model := range deviceModelList {
+		klog.Infof("======ParseByUsingRegister model %+v", model)
```
Comment (RyanZhaoXB) [inline pkg/util/parse/parse.go:379]: these log in this file are for test. please remove them.

**Review by RyanZhaoXB — CHANGES_REQUESTED** [review]
please focus on the modification about DMI in this pr. some good optimizations which are not related with DMI could be proposed in another pr.
and please remove the binary of modbus in this formal pr.

**Comment by wackxu** [comment]
I see that we update the code in the original logic code, shall we create new dir and files to keep the old mapper until the dmi is in production. cc @RyanZhaoXB

File: mappers/modbus/cmd/main.go [hunk 21]
```diff
@@ -38,21 +44,68 @@ func main() {
...
+			SockPath: c.GrpcServer.SocketPath,
+			Protocol: common.ProtocolModbus,
+		},
+	)
+	go func() {
+		_ = grpcServer.Start()
+	}()
+	klog.Infoln("grpc server start finished")
+
+	global.MqttClient = common.MqttClient{
```
[the remark this answers is not in the document]
Comment (RyanZhaoXB) [reply mappers/modbus/cmd/main.go:63]: > I see that we update the code in the original logic code, shall we create new dir and files to keep the old mapper until the dmi is in production. cc @RyanZhaoXB

@wackxu yes. i have also raised this issue before. let me help @cl2017 to make it another directory.

File: pkg/apis/dmi/v1/api.proto (hunk 2 above)
Comment (gy95) [reply pkg/apis/dmi/v1/api.proto:1]: only can be merged after v1.12.0 release?

File: mappers/modbus/device/twindata.go [hunk 22]
```diff
@@ -119,36 +123,70 @@ func TransferData(isRegisterSwap bool, isSwap bool,
...
+		klog.Errorf("twindata %s get payload failed, err: %s", td.Name, err)
+		return
+	}
+	if err = global.MqttClient.Publish(td.Topic, payload); err != nil {
 		klog.Errorf("Publish topic %v failed, err: %v", td.Topic, err)
 	}
-	klog.V(2).Infof("Get the %s value as %s", td.Name, sData)
+}
+
+func (td *TwinData) FakeReport() {
```
[the remark this answers is not in the document]
Comment (cl2017) [reply mappers/modbus/device/twindata.go]: done

File: mappers/modbus/device/twindata.go [hunk 23]
```diff
@@ -119,36 +123,70 @@ func TransferData(isRegisterSwap bool, isSwap bool,
...
 		}
 	}
-	if err = globals.MqttClient.Publish(td.Topic, payload); err != nil {
+	klog.V(2).Infof("Get the %s value as %s", td.Name, sData)
+	return payload, nil
+}
+
+// Run timer function.
+func (td *TwinData) Run() {
+	td.FakeReport()
```
[the remark this answers is not in the document]
Comment (cl2017) [reply mappers/modbus/device/twindata.go]: done

File: pkg/grpcserver/device.go (hunk 4 above)
Comment (cl2017) [reply pkg/grpcserver/device.go]: done

File: pkg/grpcserver/device.go [hunk 24]
```diff
@@ -0,0 +1,198 @@
...
+	deviceInstance, err := parse.ParseDeviceFromGrpc(device, &model)
+	if err != nil {
+		return nil, fmt.Errorf("parse device %s instance failed, err: %s", device.Name, err)
+	}
+	deviceInstance.PProtocol = protocol
+
+	s.devPanel.UpdateDev(&model, deviceInstance, &protocol)
+
+	// TODO need edgecore publish?
+	// publish device twin to mqtt
```
[the remark this answers is not in the document]
Comment (cl2017) [reply pkg/grpcserver/device.go]: done

File: pkg/util/grpcclient/register.go (hunk 7 above)
Comment (cl2017) [reply pkg/util/grpcclient/register.go]: done

File: pkg/util/grpcclient/register.go (hunk 8 above)
Comment (cl2017) [reply pkg/util/grpcclient/register.go]: done

File: pkg/util/grpcclient/register.go [hunk 25]
```diff
@@ -0,0 +1,66 @@
...
+			ApiVersion: cfg.Common.APIVersion,
+			Protocol:   cfg.Common.Protocol,
+			Address:    []byte(cfg.GrpcServer.SocketPath),
+			State:      common.DEVSTOK,
+		},
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+	fmt.Println("======send grpc to edgecore for mapperRegister finished")
```
[the remark this answers is not in the document]
Comment (cl2017) [reply pkg/util/grpcclient/register.go]: done

File: pkg/util/grpcclient/register.go (hunk 9 above)
Comment (cl2017) [reply pkg/util/grpcclient/register.go]: done

File: pkg/util/grpcclient/report.go (hunk 10 above)
Comment (cl2017) [reply pkg/util/grpcclient/report.go:32]: done

File: pkg/util/parse/grpc.go (hunk 11 above)
Comment (cl2017) [reply pkg/util/parse/grpc.go]: done

File: pkg/util/parse/grpc.go (hunk 15 above)
Comment (cl2017) [reply pkg/util/parse/grpc.go]: done

File: pkg/util/parse/grpc.go (hunk 16 above)
Comment (cl2017) [reply pkg/util/parse/grpc.go]: done

File: pkg/util/parse/grpc.go (hunk 17 above)
Comment (cl2017) [reply pkg/util/parse/grpc.go]: done

File: pkg/util/parse/grpc.go (hunk 18 above)
Comment (cl2017) [reply pkg/util/parse/grpc.go]: done

File: pkg/util/parse/parse.go (hunk 19 above)
Comment (cl2017) [reply pkg/util/parse/parse.go]: done

File: pkg/util/parse/parse.go (hunk 20 above)
Comment (cl2017) [reply pkg/util/parse/parse.go]: done

File: mappers/modbus/device/devstatus.go [hunk 26]
```diff
@@ -17,16 +17,16 @@ limitations under the License.
 package device
 
 import (
-	"k8s.io/klog/v2"
+	"github.com/kubeedge/mappers-go/pkg/common"
+	"github.com/kubeedge/mappers-go/pkg/driver/modbus"
+	"github.com/kubeedge/mappers-go/pkg/global"
```
[the remark this answers is not in the document]
Comment (cl2017) [reply mappers/modbus/device/devstatus.go:22]: global is used for DMI

File: pkg/httpserver/server.go [hunk 27]
```diff
@@ -0,0 +1,133 @@
+package httpserver
```
[the remark this answers is not in the document]
Comment (cl2017) [reply pkg/httpserver/server.go:1]: todo remove them

File: mappers/modbus/cmd/main.go [hunk 28]
```diff
@@ -38,21 +44,68 @@ func main() {
...
+		klog.Infoln("======dev init mode is not register, will register to edgecore")
+		// TODO health check
+		if _, _, err = grpcclient.RegisterMapper(&c, false); err != nil {
+			klog.Fatal(err)
+		}
+		klog.Infoln("registerMapper finished")
+	}
+
+	panel.DevStart()
+	_ = httpserver.StartHTTPServer(c.HttpServer.Host)
```
[the remark this answers is not in the document]
Comment (cl2017) [reply mappers/modbus/cmd/main.go:109]: done

File: mappers/modbus/config.yaml [hunk 29]
```diff
@@ -4,3 +4,9 @@ mqtt:
   password: ""
   certification: ""
 configmap: /opt/kubeedge/deviceProfile.json
+metaserver:
```
[the remark this answers is not in the document]
Comment (cl2017) [reply mappers/modbus/config.yaml]: done

File: mappers/modbus/device/devstatus.go [hunk 30]
```diff
@@ -17,16 +17,16 @@ limitations under the License.
...
 
-	"github.com/kubeedge/mappers-go/mappers/common"
-	"github.com/kubeedge/mappers-go/mappers/modbus/driver"
-	"github.com/kubeedge/mappers-go/mappers/modbus/globals"
+	"k8s.io/klog/v2"
```
[the remark this answers is not in the document]
Comment (cl2017) [reply mappers/modbus/device/devstatus.go]: done

File: mappers/modbus/device/twindata.go [hunk 31]
```diff
@@ -119,36 +123,70 @@ func TransferData(isRegisterSwap bool, isSwap bool,
 	}
 }
 
-// Run timer function.
-func (td *TwinData) Run() {
+func (td *TwinData) GetPayload() ([]byte, error) {
 	var err error
+	klog.Infof("td: %+v", td)
```
[the remark this answers is not in the document]
Comment (cl2017) [reply mappers/modbus/device/twindata.go]: done

File: pkg/util/parse/grpc.go (hunk 12 above)
Comment (cl2017) [reply pkg/util/parse/grpc.go]: done

File: pkg/util/parse/grpc.go (hunk 13 above)
Comment (cl2017) [reply pkg/util/parse/grpc.go]: done

File: pkg/util/grpcclient/register.go (hunk 6 above)
Comment (cl2017) [reply pkg/util/grpcclient/register.go]: done

File: pkg/grpcserver/upstream.go (hunk 5 above)
Comment (cl2017) [reply pkg/grpcserver/upstream.go]: done

File: pkg/grpcserver/server.go [hunk 32]
```diff
@@ -0,0 +1,81 @@
+package grpcserver
+
+import (
+	"fmt"
+	"net"
+	"os"
+
+	modbusdevice "github.com/kubeedge/mappers-go/mappers/modbus/device"
```
[the remark this answers is not in the document]
Comment (cl2017) [reply pkg/grpcserver/server.go]: done

File: pkg/driver/modbus/client.go (hunk 3 above)
Comment (cl2017) [reply pkg/driver/modbus/client.go]: done

**Comment by fisherxu** [comment]
@cl2017 Please fix the failed ci tests, thanks.

**Pull request closed without merging**